lib/plugins/aws/package/lib/mergeIamTemplates.js

```diff
@@ -39,8 +39,12 @@ module.exports = {
 
     const iamRole = _.get(this.serverless.service.provider.iam, 'role', {});
 
-    // resolve early if provider level role is provided
-    if (typeof iamRole === 'string') {
+    // resolve early if provider level role is provided as string or CF intrinsic function
+    if (
+      typeof iamRole === 'string' ||
+      (_.isObject(iamRole) &&
+        ['Fn::Sub', 'Fn::ImportValue', 'Fn::GetAtt'].some((key) => key in iamRole))
+    ) {
```

Comment:
Shouldn't this be simply: `if (iamRole) return;` ?

Reply:
Not really, because we support notation like this:

```
provider:
  name: aws
  iam:
    role:
      statements:
        - Effect: 'Allow'
          Action:
            - 's3:ListBucket'
          Resource:
            Fn::Join:
              - ''
              - - 'arn:aws:s3:::'
                - Ref: ServerlessDeploymentBucket
```

where `role` is also an object.

Reply:
Indeed I forgot, it's either settings for _default_ role, or ARN of external role to be used.

Still in such case I think it'll be good to patch it in `getCustomExecutionRole` and in other places refer to this method explicitly to check whether it's custom or default role.

This check is not very straightforward, so I think it's best if it's implemented in one place to which we refer. WDYT?

Reply:
Good suggestions :+1: I've went with a bit different approach of introducing a helper method used in `getCustomExecutionRole` and `mergeIamTemplates`. The reason for that was that `getCustomExecutionRole` is specific to getting a role for a function and `mergeIamTemplates` operates in the context of whole service. Let me know what do you think :+1: